Add Design Doc

wadells opened the pull request:
The QA Interview challenge specified:

> Before writing the actual code, we ask you to create a small design and test plan document in Google Docs and share with the team. This document should consist of key trade-offs and key design approaches. Please avoid writing an overly detailed design document, use this document to make sure the team could provide design feedback and demonstrate that you've investigated the problem space to provide a reasonable design.

This is my first draft of the design document, focusing mostly on what is in & out of scope and some minimal technology choices.

Feedback I'd love: 
* Are there specific items that are in scope that should be out of scope?
* Are there specific items that I didn't mention that you really want to see included in the design?
* Anything else you think is worthwile.

wadells committed:
Add 1st draft of design  doc.

Per the interview document, the Gravitational team wanted a
*brief* design doc covering what is in and out of scope and why.

This is my first swing at it after a couple questions with Sasha in
slack. I expect revisions.

wadells commented:
The blip of a change was a "wip" commit that I forgot to nix before making the PR.  I caught it in my own review.  No actual code change, just a cleaner commit history for y'all.

bernardjkim reviewed (approved):
If you follow a particular style guide for python, please include a link so we can use it as reference.
Design doc looks good to me.

wadells committed:
Add note about preferred python style.

bernardjkim requested that I specify what coding guidelines I choose to
follow.  I'm not a big fan of manual style enforcement as humans are
rather fallible about following rules consistently.  Thus I've chosen my
favorite automated tool and settings -- flake8 with defaults except max
line length -- and specified that as the preferred style for the
project.

Typically I'd have this be auto enforced by a CI lint/static analysis
build on every PR.  I'm not gonna go quite that far for this repo yet
though, as my editor is set up to auto enforce, and I don't know that
this repo will ever have mutiple developers contributing (thus
necessitating CI to enforce consistency across devs).

wadells commented:
Thanks Bernard!  I added a note about the style/linting I follow for python (flake8 which is a slight extension over pep8).  Let me know if you have further questions about my philosophy on code style or automated enforcement.
